iothub/device/src/Transport/AmqpClientConnectionPool.cs

```diff
@@ -0,0 +1,155 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.Azure.Devices.Client.Transport
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Text;
+    using System.IO;
+    using System.Collections.Concurrent;
+    using Microsoft.Azure.Devices.Shared;
+
+    delegate void RemoveClientConnectionFromPool(DeviceClientEndpointIdentity deviceClientEndpointIdentity);
+
+    /// <summary>
+    /// 
+    /// </summary>
+    internal class AmqpClientConnectionPool
+    {
+        ConcurrentDictionary<DeviceClientEndpointIdentity, AmqpClientConnection> iotHubSasPool;
+        ConcurrentDictionary<DeviceClientEndpointIdentity ,AmqpClientConnection> singleSasPool;
+        ConcurrentDictionary<DeviceClientEndpointIdentity ,AmqpClientConnection> sasMuxPool;
+        ConcurrentDictionary<DeviceClientEndpointIdentity ,AmqpClientConnection> x509Pool;
+
+        /// <summary>
+        /// 
+        /// </summary>
+        internal AmqpClientConnectionPool()
+        {
+            if (Logging.IsEnabled) Logging.Enter(this, $"{nameof(AmqpClientConnectionPool)}");
+
+            singleSasPool = new ConcurrentDictionary<DeviceClientEndpointIdentity, AmqpClientConnection>();
+            x509Pool = new ConcurrentDictionary<DeviceClientEndpointIdentity, AmqpClientConnection>();
+            sasMuxPool = new ConcurrentDictionary<DeviceClientEndpointIdentity, AmqpClientConnection>();
+            iotHubSasPool = new ConcurrentDictionary<DeviceClientEndpointIdentity, AmqpClientConnection>();
+        }
+
+        internal AmqpClientConnection GetClientConnection(DeviceClientEndpointIdentity deviceClientEndpointIdentity)
```

Comment:
Is it a good idea to throw exception if there is no connection found?

Reply:
Yes, if there is no connection created throwing an exception sounds like a good idea for me. 
I will change the exception to ArgumentOutOfRangeException with a message though.

---
In reply to: [251627827](https://github.com/Azure/azure-iot-sdk-csharp/pull/770#discussion_r251627827) [](ancestors = 251627827)